cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -145,45 +156,153 @@ func runApply(flags *applyFlags, userVersion string) error {
 		}
 	}
 
-	waiter := getWaiter(flags.dryRun, client)
-
-	// Use a prepuller implementation based on creating DaemonSets
-	// and block until all DaemonSets are ready; then we know for sure that all control plane images are cached locally
-	klog.V(1).Infoln("[upgrade/apply] creating prepuller")
-	prepuller := upgrade.NewDaemonSetPrepuller(client, waiter, &cfg.ClusterConfiguration)
-	componentsToPrepull := constants.ControlPlaneComponents
-	if cfg.Etcd.External == nil && flags.etcdUpgrade {
-		componentsToPrepull = append(componentsToPrepull, constants.Etcd)
-	}
-	if err := upgrade.PrepullImagesInParallel(prepuller, flags.imagePullTimeout, componentsToPrepull); err != nil {
-		return errors.Wrap(err, "[upgrade/prepull] Failed prepulled the images for the control plane components error")
-	}
-
 	// Now; perform the upgrade procedure
 	klog.V(1).Infoln("[upgrade/apply] performing upgrade")
-	if err := PerformControlPlaneUpgrade(flags, client, waiter, cfg); err != nil {
+	if err := PerformControlPlaneUpgrade(flags, applyData.Client(), applyData.Waiter(), applyData.Cfg()); err != nil {
```

Comment:
Should be `PerformControlPlaneUpgrade(flags, applyData)` or even `PerformControlPlaneUpgrade(applyData)`

Reply:
I did not because I was trying not to change any of the functions. but I would agree this is cleaner. will do.

Reply:
just means updating a long of functions. but is not hard.